### PR TITLE
64bit MMAL support

### PR DESCRIPTION
**Pull request by 6by9**
This has only had the briefest of testing, but raspistill is capturing JPEGs OK.

### CODE REVIEW

**Comment by 6by9** [comment]
Posted for basic testing.
I'm aware of an issue that the client_context handles aren't released on port disable, so not in a state that can be merged. I also haven't checked how the firmware build copes with the header changes, so there will need to be a follow up there.

**Comment by 6by9** [comment]
Force pushed which hopefully resolves the leaks.

The top patch adds debug that on every component destroy call dumps out debug of how many client or component handles are still mapped. On a clean application exit these should always be 0. This patch is not expected to be merged into the final tree.

File: interface/mmal/vc/mmal_vc_client.c [hunk 1]
```diff
@@ -477,30 +676,36 @@ static VCHIQ_STATUS_T mmal_vc_vchiq_callback(VCHIQ_REASON_T reason,
...
          }
          vchiq_release_message(service, header);
       }
       break;
    case VCHIQ_BULK_TRANSMIT_ABORTED:
       {
          mmal_worker_buffer_from_host *msg = (mmal_worker_buffer_from_host *)context;
-         LOG_INFO("bulk tx aborted: %p, %d", msg->buffer_header.data, msg->buffer_header.length);
-         vcos_assert(msg->drvbuf.client_context->magic == MMAL_MAGIC);
+         MMAL_VC_CLIENT_BUFFER_CONTEXT_T *client_context =
```
Comment (sakaki-) [inline interface/mmal/vc/mmal_vc_client.c]: @6by9 

With `-Werror` and `-Wno-unused-but-set-variable` (the default), compile fails for me (on aarch64 Gentoo, `gcc` 9.2.0) on this line, with:

```
/var/tmp/portage/media-libs/raspberrypi-userland-1.20191025/work/raspberrypi-userland-1.20191025/interface/mmal/vc/mmal_vc_client.c:705:43: error: unused variable ‘client_context’ [-Werror=unused-variable]
  705 |          MMAL_VC_CLIENT_BUFFER_CONTEXT_T *client_context =
      |                                           ^~~~~~~~~~~~~~
```

Adding something similar to

```c
client_context = NULL;
```

at the end of the function suppresses this.

Comment (6by9) [reply interface/mmal/vc/mmal_vc_client.c]: Thanks. I was building debug builds and therefore the assert used the variable.
The other approach would be to do the call within the assert. I'll revisit it when I have a chance. 

I've been trying to tidy up the firmware side of it first as it has all the counterpart casts to convert between the IPC and active formats (even though they are actually identical in a 32bit world).

**Comment by sakaki-** [comment]
Hi @6by9,

so, on the basis of _very_ preliminary testing, `mmal` now seems usable from a 64-bit userland, for h264 decoding at any rate ^-^

I've built the `raspberrypi-userland-1.20191025-r1` package with your patch (tweaked very slightly so it builds non-debug, per comment [above](https://github.com/raspberrypi/userland/pull/586#discussion_r339145148)), and then rebuilt `ffmpeg-4.2-r1` with the `mmal` USE flag. After this I was able to successfully play videos using e.g.:

```console
demouser@pi64 ~ $ ffmpeg -f video4linux2 -input_format h264 \
  -video_size 1280x720 -framerate 30 -i /dev/video0 \
  -vcodec copy -an -f matroska test.mkv
<after a while, Ctrl-c, then...>
demouser@pi64 ~ $ ffplay -vcodec h264_mmal -i test.mkv
```
Seems to work reasonably well (and is definitely using the `mmal` codec since it outputs a number of `did not get output frame from MMAL` messages to the console at the end of the stream). Also decodes other h264 files with audio etc.

Haven't tried capture or transcoding yet.

This is on a [gentoo-on-rpi-64bit](https://github.com/sakaki-/gentoo-on-rpi-64bit) system, with `bcm2711-kernel-bis-bin-4.19.80.20191022` (and - because of the way this kernel package is set up - by default the boot firmware will also be whatever was current as of that date).

hth, sakaki

File: host_applications/linux/apps/raspicam/RaspiTexStub.c [hunk 2]
```diff
@@ -0,0 +1,240 @@
+/*
```
Comment (pelwell) [inline host_applications/linux/apps/raspicam/RaspiTexStub.c]: I wasn't expecting a new source file based on the commit message (which has a typo, BTW). A bit more explanation would be nice.

File: interface/mmal/vc/mmal_vc_msgs.h [hunk 3]
```diff
@@ -515,7 +612,7 @@ static inline void mmal_vc_buffer_header_to_msg(mmal_worker_buffer_from_host *ms
    msg->buffer_header.pts           = header->pts;
    msg->buffer_header.dts           = header->dts;
    msg->buffer_header.alloc_size    = header->alloc_size;
-   msg->buffer_header.data          = header->data;
+   msg->buffer_header.data          = (uintptr_t)header->data;
    msg->buffer_header_type_specific = *header->type;
 }
```
Comment (pelwell) [inline interface/mmal/vc/mmal_vc_msgs.h]: This whole commit is a bit TL;DR for me to have fully analysed it, but it all looks plausible.

File: interface/mmal/vc/mmal_vc_client.c [hunk 4]
```diff
@@ -453,18 +613,21 @@ static VCHIQ_STATUS_T mmal_vc_vchiq_callback(VCHIQ_REASON_T reason,
          if (msg_hdr->msgid == MMAL_WORKER_BUFFER_TO_HOST)
          {
             mmal_worker_buffer_from_host *msg = (mmal_worker_buffer_from_host *)msg_hdr;
-            vcos_assert(msg->drvbuf.client_context->magic == MMAL_MAGIC);
-            msg->drvbuf.client_context->callback(msg);
-            LOG_TRACE("bulk rx done: %p, %d", msg->buffer_header.data, msg->buffer_header.length);
+            MMAL_VC_CLIENT_BUFFER_CONTEXT_T *client_context = mmal_vc_lookup_client_context(msg->drvbuf.client_context);
```
Comment (popcornmix) [inline interface/mmal/vc/mmal_vc_client.c]: is indentation screwy here?

Comment (pelwell) [reply interface/mmal/vc/mmal_vc_client.c]: It looks OK to me...

File: interface/mmal/vc/mmal_vc_client.c [hunk 5]
```diff
@@ -82,6 +83,143 @@ struct MMAL_CLIENT_T
...
+   if (vcos_verify(i != MAX_CLIENT_CONTEXTS))
+   {
+      client_context_pool.contexts[i].index = i;
+      client_context_pool.contexts[i].ctx = context;
+      client_context_pool.contexts[i].inuse = 1;
+   }
+   vcos_mutex_unlock(&client_context_pool.lock);
+
+   return i;
+}
```
Comment (popcornmix) [inline interface/mmal/vc/mmal_vc_client.c]: Has this been tested when MAX_CLIENT_CONTEXTS has been reached?
It looks like this returns MAX_CLIENT_CONTEXTS, so mmal_vc_lookup_client_context returns NULL.
But that doesn't seem to be protected against (e.g. mmal_vc_port_by_number dereferences null pointer)

File: interface/mmal/vc/mmal_vc_msgs.h (hunk 3 above)
Comment (6by9) [reply interface/mmal/vc/mmal_vc_msgs.h]: Sorry, I had started with bits of it as some individual commits, but hadn't been careful enough about squashing fixups into the right commit, so squahsed the whole lot into one.
It's mainly using a lookup table for client contexts and component contexts - I can try splitting the various bits out into independent commits if you really want it. There's then the changes to the RPC structures as well with some casts.

Annoyingly the contexts get allocated (and freed) from mmal_vc_api.c, but all the actual lookups are in mmal_vc_client, hence exporting only some of the functions.

File: interface/mmal/vc/mmal_vc_client.c (hunk 5 above)
Comment (6by9) [reply interface/mmal/vc/mmal_vc_client.c]: No, the failure mode is currently going to be nasty. The limits are set so that they are highly unlikely to ever be hit, but the alloc does need to fail gracefully and then pass the failure back up.

**Comment by 6by9** [comment]
I do have the firmware side compiling too, but I'll hold fire on pushing that one so that I can cherry-pick the actual changes from here rather than having a mismash.
The firmware changes aren't required for any functionality, only to bring the RPC structures into sync.

File: interface/mmal/vc/mmal_vc_msgs.h (hunk 3 above)
Comment (pelwell) [reply interface/mmal/vc/mmal_vc_msgs.h]: Pasting that into the commit message is probably enough.

**Comment by popcornmix** [comment]
Does firmware side require a rebuild?
A sanity check with (32-bit) kodi seemed to work initially.
It got a bit chattier (mmalipc) but it hung when stopping video playback.
An interlaced video failed to start.

**Comment by 6by9** [comment]
As above, the firmware shouldn't need changing at all.
I hadn't remembered adding additional logging except for the debug commit on the top to try and head off leaks.
I'll try Kodi for myself when I'm back in the office. I suspect that is making greater use of events than my tests.

**Comment by popcornmix** [comment]
```
mmalipc: mmal_vc_dump_client_components: mmal_vc_dump_client_components: Entry 1 in use for context 0x74c8d340
mmalipc: mmal_vc_dump_client_components: mmal_vc_dump_client_components: Entry 2 in use for context 0x63618d40
mmalipc: mmal_vc_dump_client_components: mmal_vc_dump_client_components: 2 entries in use
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 1 in use for context 0x63408378
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 2 in use for context 0x63614104
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 3 in use for context 0x63614248
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 4 in use for context 0x63613fc0
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 5 in use for context 0x63613e7c
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 6 in use for context 0x63613d38
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 7 in use for context 0x63613bf4
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 8 in use for context 0x63613ab0
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 9 in use for context 0x6361396c
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 10 in use for context 0x63613828
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 11 in use for context 0x636136e4
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 12 in use for context 0x636135a0
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 13 in use for context 0x6361345c
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 14 in use for context 0x63613318
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 15 in use for context 0x636131d4
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 16 in use for context 0x63613090
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 17 in use for context 0x63612f4c
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 18 in use for context 0x63612e08
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 19 in use for context 0x63612cc4
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 20 in use for context 0x63612b80
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 21 in use for context 0x63612a3c
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 22 in use for context 0x636128f8
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 23 in use for context 0x636127b4
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 24 in use for context 0x63612670
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 25 in use for context 0x6361252c
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 26 in use for context 0x636123e8
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 27 in use for context 0x636122a4
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 28 in use for context 0x6365ad18
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: 28 entries in use
```

File: interface/mmal/vc/mmal_vc_msgs.h (hunk 3 above)
Comment (6by9) [reply interface/mmal/vc/mmal_vc_msgs.h]: OK, I'll update the commit message with more detail of what exactly has been changed and why.

**Comment by 6by9** [comment]
All from https://github.com/raspberrypi/userland/pull/586/commits/605fe8ebe9fc7a6ab1eefa6f38a445e1d6bb9a84 "Debug for context and component mappings".
Sorry, forgot to add a "Do not merge" into the commit message, but wasn't worrying as I'd tagged the PR as RFC.

I'm just hoping that on a clean exit you end up with 0 entries in use.

**Comment by popcornmix** [comment]
```
kodi-rbpi: /home/dom/xbmc_holder/xbmc/xbmc/cores/VideoPlayer/DVDCodecs/Video/MMALCodec.cpp:223: void MMAL::CMMALVideo::dec_output_port_cb(MMAL_PORT_T*, MMAL_BUFFER_HEADER_T*): Assertion `omvb' failed.
```
Got one of these, which suggests the callback pointer I passed in wasn't returned as expected.

**Comment by popcornmix** [comment]
```
assertion failure:/home/dom/projects/vc4/interface/mmal/vc/mmal_vc_client.c:212:mmal_vc_lookup_client_context():client_context_pool.contexts[index].inuse
```

**Comment by 6by9** [comment]
OK, I'll have a look at what kodi is doing when I'm back in the office. 32bit should have been almost a nop, but obviously not. And I'm very glad I didn't make this all a special path only for 64bit as it wouldn't have shaken out these issues.

**Comment by 6by9** [comment]
I've fixed up what I believe was the issue.  Two extra patches pushed which I'll squash when confirmed to work. I've only tested on 32bit at present, so will try 64bit when I get some more time.

**Comment by sakaki-** [comment]
HI @6by9,

FYI, I just [merged this PR](https://www.raspberrypi.org/forums/viewtopic.php?p=1572720#p1572720) for the recent v1.5.2 release of [gentoo-on-rpi-64bit](https://github.com/sakaki-/gentoo-on-rpi-64bit). Appreciate it's been tagged with RFC status, but having it out there should hopefully elicit some "C" ^-^

Best, sakaki

**Comment by popcornmix** [comment]
@6by9 sorry - missed your updated commits. Just tested again, kodi-mmal, pi4 with H264 and MPEG2 on 32-bit userland, with 32-bit kernel seems fine now.

With 64-bit kernel (and so the switch from vcsm to vcsm-cma) it is mostly functional.
Software decode seems fine. H264 has a lot of green corruption. Possibly caching coherency issues, possibly lost encoded data bytes.

Example output after playing H264. 
```
mmalipc: mmal_vc_dump_client_components: mmal_vc_dump_client_components: Entry 1 in use for context 0xef9f75d0
mmalipc: mmal_vc_dump_client_components: mmal_vc_dump_client_components: Entry 2 in use for context 0xe52244f0
mmalipc: mmal_vc_dump_client_components: mmal_vc_dump_client_components: 2 entries in use
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 1 in use for context 0xef9f8bd8
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 2 in use for context 0xe520a4d0
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 3 in use for context 0xe520ac68
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 4 in use for context 0xe520ab24
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 5 in use for context 0xe520a9e0
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 6 in use for context 0xe520a89c
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 7 in use for context 0xe520a758
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 8 in use for context 0xe520a614
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 9 in use for context 0xe520a38c
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 10 in use for context 0xe520a248
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 11 in use for context 0xe520a104
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 12 in use for context 0xe5209fc0
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 13 in use for context 0xe5209e7c
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 14 in use for context 0xe5209d38
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 15 in use for context 0xe5209bf4
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 16 in use for context 0xe5209ab0
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 17 in use for context 0xe520996c
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 18 in use for context 0xe5209828
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 19 in use for context 0xe52096e4
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 20 in use for context 0xe52095a0
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 21 in use for context 0xe520945c
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 22 in use for context 0xe5209318
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 23 in use for context 0xe52091d4
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 24 in use for context 0xe5209090
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 25 in use for context 0xe5208f4c
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 26 in use for context 0xe5208e08
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 27 in use for context 0xe5208cc4
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 28 in use for context 0xef98af18
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: 28 entries in use
mmalipc: mmal_vc_dump_client_components: mmal_vc_dump_client_components: Entry 0 in use for context 0xedb09460
mmalipc: mmal_vc_dump_client_components: mmal_vc_dump_client_components: Entry 2 in use for context 0xe52244f0
mmalipc: mmal_vc_dump_client_components: mmal_vc_dump_client_components: 2 entries in use
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 0 in use for context 0xedb09878
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 28 in use for context 0xef98af18
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: 2 entries in use
mmalipc: mmal_vc_dump_client_components: mmal_vc_dump_client_components: Entry 2 in use for context 0xe52244f0
mmalipc: mmal_vc_dump_client_components: mmal_vc_dump_client_components: 1 entries in use
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: Entry 28 in use for context 0xef98af18
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: 1 entries in use
mmalipc: mmal_vc_dump_client_components: mmal_vc_dump_client_components: 0 entries in use
mmalipc: mmal_vc_dump_client_contexts: mmal_vc_dump_client_contexts: 0 entries in use
```

**Comment by jordanjohnp** [comment]
I'd just like to mention that I tested this on pi 4, Ubuntu 18.04 64-bit (using `https://github.com/TheRemote/Ubuntu-Server-raspi4-unofficial/releases'), and recording (raspivid) is working fine for me (recorded video is clean). I do see the messages similar to:

`mmalipc: mmal_vc_dump_client_components: mmal_vc_dump_client_components: 0 entries in use`

Not sure if this is the right place to provide such feedback, but thanks 6by9!

**Comment by 6by9** [comment]
The debug messages are from https://github.com/raspberrypi/userland/pull/586/commits/605fe8ebe9fc7a6ab1eefa6f38a445e1d6bb9a84, which is only there for test purposes. Getting "0 entries in use" means that everything is being cleaned up.

The suspicion is that there are still a couple of corner cases in there that aren't behaving, but I'm tempted to tidy up the patches and merge this. It doesn't harm 32bit, and seems to support many 64 bit use cases.

**Comment by popcornmix** [comment]
@6by9 I'm okay with getting this merged. I didn't spot any regressions and it made some 64-bit use cases work better.

**Comment by 6by9** [comment]
Sorry, I've noticed that I've got a couple of minor updates on my working tree.

Three main things:
- condensing ```index``` and ```inuse``` to being ```unsigned int inuse:1; unsigned int index:31``` as inuse is effectively a bool, and index will never exceed ```MAX_CLIENT_CONTEXTS``` or ```MAX_COMPONENT_HANDLES```
- Having saved that memory, increasing ```MAX_COMPONENT_HANDLES``` to 128 (was 64), and ```MAX_CLIENT_CONTEXTS``` to 512 (was 128). At 8 bytes an entry now those tables are still only 5kB in size.
- Handle a client context lookup for buffers with ```has_reference``` set. I'm not 100% clear on the use case and whether it is actively used, but I think I understand it well enough. (Things like a splitter component that is just passing out a reference to the input buffer to each output. The output buffers get a reference set against the input buffer to manage lifetime).

I'll drop/disable the debug patches, and add those changes in.

**Comment by 6by9** [comment]
I've found another corner case. The MMAL_EVENT_FORMAT_CHANGED event buffer includes the MMAL_ES_FORMAT_T and MMAL_ES_SPECIFIC_FORMAT_T and MMAL_ES_FORMAT_T  is one of the structures that changes size.
mmal_event_format_changed_get throws an error that the buffer size is insufficient for the message.

**Comment by 6by9** [comment]
- MMAL_EVENT_FORMAT_CHANGED fixed (see https://github.com/6by9/userland/commit/7803a86ac312c0ec3054189659a7d12bb2dbc32b for the change required).
- Debug logging dropped.
- Changes squashed into one. 

I've had mmalplay working fine on Gentoo, so I believe that that is resolved.

Removing RFC tag. Any further comments, or happy to merge?
